Add FERC EQR GCS Archiver

zschira opened the pull request:
<!--
Resources:
* contributing guidelines: https://catalystcoop-pudl.readthedocs.io/en/latest/CONTRIBUTING.html
* code of conduct: https://catalystcoop-pudl.readthedocs.io/en/latest/code_of_conduct.html
-->
# Overview

This PR creates an EQR archiver that can be used with the `fsspec` storage backend. We had a draft EQR archiver in this repo previously, but had not actually used this archiver as we didn't have a plan to deal with the large scale of EQR data. This new version has actually been used to take a snapshot of raw EQR data.

Running this archiver requires using the `fsspec` backend as EQR data is much too large for Zenodo deposition size limits. This can be done with the following command:

```
pudl_archiver --datasets ferceqr --deposition-path {fsspec_compatible_path} --depositor fsspec --initialize
```

The `--initialize` argument is required if `fsspec_compatible_path` does not point to a path with a complete existing archive.

zaneselvans commented on src/pudl_archiver/archivers/ferc/ferceqr.py:
```diff
@@ -9,79 +8,41 @@
         """Download FERC EQR resources."""
-        # Get non-transaction data (pre-2013)
-        # Skip all pre-2013 data until access to FTP server is figured out
-        """
```
We got [a single static archive of this up on Zenodo](https://zenodo.org/records/10086109) if we ever need it. I doubt they will ever touch those files again... and they seem to have removed the links from [the obvious location](https://www.ferc.gov/industries-data/electric/power-sales-and-markets/electric-quarterly-reports-eqr/2nd-quarter-2013-0)

zaneselvans commented on src/pudl_archiver/archivers/ferc/ferceqr.py:
```diff
@@ -9,79 +8,41 @@
-        """Download a quarter of 2014-present data."""
-        url = f"https://eqrreportviewer.ferc.gov/DownloadRepositoryProd/BulkNew/CSV/CSV_{year}_Q{quarter}.zip"
+        """Download a quarter of 2013-present data."""
+        url = f"https://eqrreportviewer.ferc.gov/DownloadRepositoryProd/7D0F99CBC5C744969A7A9A5F4BA5612ED77CB30C09F6425BB9C3D417EFBE01C20C6C7A6DE7D0446881A7639F0FDC8FE1/BulkNew/CSV/CSV_{year}_Q{quarter}.zip"
```
Were you able to re-use this base URL throughout all of the different archiving sessions? Or did you have to update it? If you had to update it, was it every time? Every day? Every week?

zschira replied:
Hmm interesting, I used this same URL every time I ran the archiver, but I just tried using it again and I'm getting a 404 as well, so I assume it changed very recently. Unfortunately, this makes me think we need to just pull `playwright` in and start getting the URL's dynamically.

zschira commented on src/pudl_archiver/archivers/validate.py:
```diff
@@ -45,14 +45,17 @@ def validate_filetype(
         success=_validate_file_type(path, BytesIO(path.read_bytes())),
         notes=[path.name],
     )
+    if not validation.success:
```
This will result in an unrecoverable error. I think making it do a retry here would be a more involved update. In general this is a somewhat brute force way of making this fail fast. Given that we have other file level validations that we do, we might want to come up with a more standard approach where we can decide which of those should/should not fail immediately.

As for how this would impact other archivers, the difference would be that when they run they would now raise this exception immediately upon encountering an "invalid filetype", whereas normally the archiver will just keep running as if nothing happened, then output an error at the end when it actually checks the results of all the tests.

Maybe a better approach for now would be to just log a warning when a test fails, so the rest of the archiver will continue running as normally, but there will at least be an immediate acknowledgment that something is wrong, which can still be useful for a very slow archiver like EQR?

zaneselvans replied:
In general it's useful to have the archive continue even when there are some validation issues since many of them have false positives, and it's nice to be able to see all of the issues that came up in one place, so in many cases encountering an issue and continuing on, or not checking for issues until the end makes sense.

But then there are some cases in which we can identify a catastrophic failure that will make the archive unusable early (e.g. a corrupted file) and also potentially attempt to fix it (retry the download to see if it was a network issue). Being able to detect and fix this kind of problem early if possible would be helpful to all of the archivers.

And then there's also a difference between the draft Zenodo archives where something being broken like a corrupted file isn't a disaster because it's just a draft and nobody can see it until we approve it (which we hopefully would never do if there was a corrupted file) and the kind of snapshot we're doing here with EQR, where anything that's uploaded is immediately live and potentially being used as a source for downstream processing, so we need to be more conservative about whether we upload something corrupted.

In the case of these non-versioned snapshots, there's also already a version of the file that's being uploaded available, even if it's not as recent, so not uploading a corrupted file isn't a big failure -- it just means that quarter of data is a little bit stale, and presumably others will notice the issue and FERC will fix the file, and it can be updated next time.

In the case of the EQR, one behavior that might make sense is:
- Validate the zipfile on download.
- If it is not valid, retry the download N times.
- If the retries fail, log it and move on to the next quarter of data.
- Upload the file (such that it doesn't clobber the existing file until it's know to be valid) and validate the checksum.
- If the checksum fails, retry N times.
- If the retries fail, log it and move on to the next quarter of data.

zaneselvans replied:
Are there scenarios with the other datasets where this set of behaviors would not be desirable? Uploading a corrupted file to Zenodo also isn't going to be helpful / informative. Moving on after some retries would let us gather all the other interesting information about the update, and also result in an obvious issue with the new archive -- a whole (corrupted) file is missing.

I guess one downside is we wouldn't have that specific file to look at, which might make debugging more difficult (e.g. if it was "we got some HTML from an error page instead of a Zipfile" and not "this Zipfile is corrupted.").

zschira commented on src/pudl_archiver/archivers/ferc/ferceqr.py:
```diff
@@ -9,79 +8,41 @@
-        url = f"https://eqrreportviewer.ferc.gov/DownloadRepositoryProd/BulkNew/CSV/CSV_{year}_Q{quarter}.zip"
+        """Download a quarter of 2013-present data."""
+        url = f"https://eqrreportviewer.ferc.gov/DownloadRepositoryProd/7D0F99CBC5C744969A7A9A5F4BA5612ED77CB30C09F6425BB9C3D417EFBE01C20C6C7A6DE7D0446881A7639F0FDC8FE1/BulkNew/CSV/CSV_{year}_Q{quarter}.zip"
         download_path = self.download_directory / f"ferceqr-{year}-Q{quarter}.zip"
```
Took your suggestion, and I'll try to manually change the filenames.

zschira replied:
Thinking about it, I could write a script to change all the filenames based on a pattern, but maybe it would be easier to just leave it as is, and just let the archiver rename all the files next time we run it?

zschira commented on src/pudl_archiver/archivers/ferc/ferceqr.py:
```diff
@@ -9,79 +8,41 @@
             for quarter in range(1, 5):
-                if quarter < 3:
+                if (quarter < 3 and year == 2013) or (quarter > 2 and year == 2025):
                     continue
```
Based on the comment above about the hash changing in the URL, we might just need to pull playwright in meaning that we will get the full list of URL's dynamically anyway.

zaneselvans replied:
Unfortunately, that seems to be the case.

zaneselvans commented on src/pudl_archiver/archivers/ferc/ferceqr.py:
````diff
@@ -9,79 +8,41 @@
+    following command:
+
+    ```
+    pudl_archiver --datasets ferceqr --deposition-path gs://archives.catalyst.coop/eqr --depositor fsspec
````
Okay I just updated all the paths to use `ferceqr-YYYYqQ.zip` for the filenames. Making a small PR into the EQR repo now to make it work with these changes.

zschira commented:
@zaneselvans I've reworked this to use `playwright`, and it's been working well in local testing. I also removed the fail fast update I made to the filetype validation, because I plan to handle that in a more thought out way in a follow up PR (same with adding the updates to enable the GCS backend from a github runner).
